### PR TITLE
Option ``--on-server-lost``

### LINKED ISSUE #192: Continue on server connection lost
In the current version, the worker is immediately terminated when the connection to the server is lost.
Add a **worker** startup option  ``--on-server-lost=X`` where X means:
* ``stop`` = the current behavior
* ``finish-running`` = do not start any new task, but finish running tasks and then terminates
* ``finish-all`` = finish all assigned tasks and then terminates

### PR DESCRIPTION
**Pull request by spirali**
Closes #192

### CODE REVIEW

File: crates/tako/src/worker/rpc.rs [hunk 1]
```diff
@@ -98,13 +99,54 @@ pub async fn connect_to_server_and_authenticate(
...
+        );
+        match on_server_lost {
+            ServerLostPolicy::Stop => Ok(()),
+            ServerLostPolicy::FinishRunning => {
+                state.get_mut().drop_non_running_tasks();
+                let notify = Rc::new(Notify::new());
+                state.get_mut().worker_is_empty_notify = Some(notify.clone());
+                let is_empty = state.get().is_empty();
+                if is_empty {
+                    log::info!("No running tasks remains")
```
Comment (Kobzol) [inline crates/tako/src/worker/rpc.rs:128]: `No running tasks remain`

File: docs/deployment/worker.md [hunk 2]
```diff
@@ -146,3 +146,10 @@ If you also want to include workers that are offline (i.e. that have crashed or
 ```bash
 $ hq worker info <worker-id>
 ```
+
+## Lost connection to the server
+
+The behavior of what should happened with a worker that loot connection to the server is configured via ``hq worker start --on-server-lost=X`` where ``X`` can be:
```
Comment (Kobzol) [inline docs/deployment/worker.md:152]: `loot` -> `lost`

Let's move this before `Useful worker commands`

File: docs/deployment/worker.md [hunk 3]
```diff
@@ -146,3 +146,10 @@ If you also want to include workers that are offline (i.e. that have crashed or
...
+
+* ``stop`` (default) - The worker immediately terminates while killing or currently running tasks
```
Comment (Kobzol) [inline docs/deployment/worker.md:154]: Extra `or` (maybe you meant `all`?)

File: docs/deployment/worker.md [hunk 4]
```diff
@@ -146,3 +146,10 @@ If you also want to include workers that are offline (i.e. that have crashed or
...
+* ``finish-running`` - The worker do no start any new tasks and tries to finish already running tasks. When all tasks are finished then it terminates.
```
Comment (Kobzol) [inline docs/deployment/worker.md:155]: `do` -> `does`

File: tests/test_worker.py [hunk 5]
```diff
@@ -213,5 +215,36 @@ def test_worker_address(hq_env: HqEnv):
...
+    hq_env.check_process_exited(worker)
+
+
+def test_server_lost_finish_running_with_task(hq_env: HqEnv):
+    hq_env.start_server()
+    worker = hq_env.start_worker(args=["--on-server-lost=finish-running"])
+    hq_env.command(["submit", "--array=1-10", "--", "bash", "-c", "sleep 2"])
+    wait_for_job_state(hq_env, 1, "RUNNING")
+    hq_env.kill_server()
+    time.sleep(0.5)
```
Comment (Kobzol) [inline tests/test_worker.py:243]: Maybe we should check that the task was finished, e.g. some file could be written to disk?

File: tests/test_worker.py [hunk 6]
```diff
@@ -213,5 +215,36 @@ def test_worker_address(hq_env: HqEnv):
...
+    assert worker.poll() is None
+    time.sleep(2.0)
+    hq_env.check_process_exited(worker)
+
```
Comment (Kobzol) [inline tests/test_worker.py:247]: Could you add tests that check that:
1) non-started tasks are dropped with "finish-running"
2) all tasks are dropped when the worker is explicitly stopped

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,11 +2,16 @@
 
 ## New features
 
+### Worker configuraiton
+  * Argument for ``--on-worker-lost`` (for ``worker start`` and autoalloc) to define what should happend
```
Comment (Kobzol) [inline CHANGELOG.md:6]: `happend` -> `happen`

File: crates/hyperqueue/src/client/commands/autoalloc.rs [hunk 8]
```diff
@@ -96,6 +97,10 @@ pub struct SharedQueueOpts {
     #[clap(long, setting = clap::ArgSettings::MultipleOccurrences)]
     resource: Vec<PassThroughArgument<ArgGenericResourceDef>>,
 
+    /// Behavior when a connection to a server is lost
+    #[clap(long, default_value = "stop")]
```
[the remark this answers is not in the document]
Comment (spirali) [reply crates/hyperqueue/src/client/commands/autoalloc.rs:101]: I was thinking about it, but my idea was that "stop" is more visible outcome. I would start with "stop" and modify it after users will complain:)

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 9]
```diff
@@ -98,6 +118,7 @@ pub async fn start_hq_worker(
     let server_address = format!("{}:{}", record.host(), record.worker_port());
     log::info!("Connecting to: {}", server_address);
 
+    let on_server_lost = opts.on_server_lost.get().clone();
```
[the remark this answers is not in the document]
Comment (spirali) [reply crates/hyperqueue/src/client/commands/worker.rs:121]: Not possible as opts is later moved out

[the remark this answers is not in the document]
Comment (Kobzol) [reply crates/hyperqueue/src/client/commands/worker.rs:121]: Then maybe `on_server_lost` should be stored in `WorkerConfiguration`?

[the remark this answers is not in the document]
Comment (spirali) [reply crates/hyperqueue/src/client/commands/worker.rs:121]: Good point. It should be in WorkerConfiguration in the first place

File: docs/deployment/worker.md [hunk 10]
```diff
@@ -146,3 +146,13 @@ If you also want to include workers that are offline (i.e. that have crashed or
...
 ```
+
+## Lost connection to the server
+
+The behavior of what should happen with a worker that lost connection to the server is configured via ``hq worker start --on-server-lost=X`` where ``X`` can be:
+
+* ``stop`` - The worker immediately terminates while killing all currently running tasks
+* ``finish-running`` - The worker does not start any new tasks and tries to finish already running tasks. When all tasks are finished then it terminates.
+
+``stop`` is a default policy when worker manually started by ``hq worker start``. When a worker is started by automatic allocation, then
```
Comment (Kobzol) [inline docs/deployment/worker.md:157]:
```suggestion
``stop`` is a default policy when worker is manually started by ``hq worker start``. When a worker is started by automatic allocation, then
```

**Review by Kobzol — APPROVED** [review]

**Pull request merged**